Meta/AnonymizedMetadataBuilder.php

```diff
@@ -0,0 +1,119 @@
+
+/**
+ * Class AnonymizedMetadataBuilder.
+ * This class aims to build an AnonymizedMetadata object based on anonymize annotations defined on your entities.
+ *
+ * @author Benoit Mazière <benoit.maziere@ekino.com>
+ */
+final class AnonymizedMetadataBuilder
+{
+    /**
+     * @var EntityManager
+     */
+    private $entityManager;
+
+    /**
+     * @var Reader
+     */
+    protected $annotationReader;
+
+    public function __construct(EntityManager $entityManager, Reader $annotationReader)
+    {
+        $this->entityManager    = $entityManager;
+        $this->annotationReader = $annotationReader;
+    }
+
+    /**
+     * return AnonymizerMetadata[]
+     */
+    public function build(): array
+    {
+        $anonymizedMetadatas = [];
+        /** @var ClassMetadata[] $classMetadatas */
+        $classMetadatas = $this->entityManager->getMetadataFactory()->getAllMetadata();
+
+        foreach ($classMetadatas as $classMetadata) {
+            $anonymizedEntity     = $this->buildAnonymizedEntityAnnotations($classMetadata);
+            $anonymizedProperties = $this->buildAnonymizedPropertiesAnnotations($classMetadata);
+
+            if (!$anonymizedEntity && !empty($anonymizedProperties)) {
+                throw AnnotationException::creationError(
+                    sprintf('You tried to anonymize a property without specifying it at class level in %s. 
+                        You should add @AnonymizedEntity() in class phpdoc', $classMetadata->getName()));
+            }
+
+            if ($anonymizedEntity) {
+                $anonymizedMetadatas[] = $this->buildAnonymizedMetadata(
+                    $classMetadata, $anonymizedEntity, $anonymizedProperties
+                );
+            }
+        }
+        
+        return $anonymizedMetadatas;
+    }
+
+    private function buildAnonymizedMetadata(
+        ClassMetadata $classMetadata,
+        AnonymizedEntity $anonymizedEntity,
+        array $anonymizedProperties): AnonymizedMetadata
+    {
+        return (new AnonymizedMetadata())
+            ->setClassMetadata($classMetadata)
+            ->setAnonymizedEntity($anonymizedEntity)
+            ->setAnonymizedProperties($anonymizedProperties);
+    }
+
+    private function buildAnonymizedPropertiesAnnotations(ClassMetadata $classMetadata): array
+    {
+        $anonymizedProperties = [];
+        $properties           = $classMetadata->getFieldNames();
+
+        foreach ($properties as $property) {
+            /** @var AnonymizedProperty|null $anonymizedProperty */
+            $anonymizedProperty = $this->annotationReader->getPropertyAnnotation(
+                new \ReflectionProperty($classMetadata->getName(), $property),
+                AnonymizedProperty::class
+            );
+
+            if (\is_null($anonymizedProperty)) {
+                continue;
+            }
+
+            $anonymizedProperty->setFieldName($property)->setColumnName($classMetadata->getColumnName($property));
```

Comment:
Is it logical to alter the annotation here? 🤔 If the field name and column name are overriden, then they should not be configurable in the annotation "options". Therefore, these info should not be stored in the annotation. WDYT?

Reply:
agree (on the fact they should not be overridden at annotation level on user side) but we need those informations to validate it against doctrine metadata. DO you have another idea to do that (we still can validate that those property can't be altered by the user at annotation construct level)?